feat: used union types for data, loading, and error

ChristianAsilo opened the pull request:
- Added new actions for handling actionKey using before and after
- Removed PokemonDetails in GetPokemonAction
- Added new Async freezed file
- Modified PokemonOverviewVmFactory for waiting and error handling
- Modified PokemonOverviewPage to used union type

![analyze](https://user-images.githubusercontent.com/128013689/227855668-02727676-a95a-49c7-b6d1-e6e9743e05a2.PNG)
![format](https://user-images.githubusercontent.com/128013689/227855673-ee86c0a8-6147-43bc-9a01-a0ecfa77a9d4.PNG)


https://user-images.githubusercontent.com/128013689/227908367-ab087def-818b-47b0-86d1-a5a385cbe7a0.mp4

chrstn06 committed:
feat: Used union types for data, loading, and error
- Added new actions for handling actionKey using before and after
- Removed PokemonDetails in GetPokemonAction
- Added new Async freezed file
- Modified PokemonOverviewVmFactory for waiting and error handling
- Modified PokemonOverviewPage to used union type

shawn-ffufm commented on lib/state/action/pokemon_actions.dart:
```diff
@@ -1,30 +1,20 @@
-}
-
-///Getting of pokemon details from pokemon api
-class GetPokemonDetailsAction extends ReduxAction<AppState> {
```
why removed this one? you only need to update GetPokemonsAction right? 😱

shawn-ffufm commented on lib/state/action/actions.dart:
```diff
@@ -0,0 +1,14 @@
+import 'package:async_redux/async_redux.dart';
+import 'package:pokedex_async_redux/state/app_state.dart';
+
+abstract class LoadingAction extends ReduxAction<AppState> {
```
you forgot to add a concise documentation 😝

remember, this is mandtoy right? 😉

shawn-ffufm commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -8,22 +10,28 @@ class PokemonOverviewPage extends StatelessWidget {
+          ),
+        ),
+        loading: () => const Center(child: CircularProgressIndicator()),
+        error: (errorMessage) => AlertDialog(
```
this seems not the proper way of calling the alert dialog 🙂

https://api.flutter.dev/flutter/material/AlertDialog-class.html

shawn-ffufm commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -8,22 +10,28 @@ class PokemonOverviewPage extends StatelessWidget {
+            itemBuilder: (_, index) {
+              final pokemon = data[index];
+              return PokemonCard(
+                pokemon: pokemon,
```
no need for extra trailing comma here 🙂

shawn-ffufm commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -8,22 +10,28 @@ class PokemonOverviewPage extends StatelessWidget {
-              pokemon: pokemon,
-            );
-          },
+      home: pokemons.when(
```
have this for the body property not home 🙂

shawn-ffufm reviewed (changes requested):
see pr feedback 🙂

chrstn06 committed:
feat: Used union types for data, loading, and error
(pr feedback)
- Modified AppState to accept PokemonDetails
- Added documentation in LoadingAction
- Removed extra trailing commas
- Modified pokemons.when into body property
- Added snackbar message
- Used PostFrame to display error

shawn-ffufm commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -8,24 +10,38 @@ class PokemonOverviewPage extends StatelessWidget {
 
   @override
   Widget build(BuildContext context) {
+    void _showErrorMessageSnackbar(String? errorMessage) {
```
tho this is allowed, please don't make it a practice to create a function inside a function unless it's super needed 🙂

have this outside then make it private instead 🙂

ChristianAsilo replied:
noted po sir

shawn-ffufm commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -8,24 +10,38 @@ class PokemonOverviewPage extends StatelessWidget {
   Widget build(BuildContext context) {
+    void _showErrorMessageSnackbar(String? errorMessage) {
+      final SnackBar snackBar = SnackBar(
+        content: Text(errorMessage!),
```
use null coalescing and use emptyString constant as its fallback value instead of using non-null enforce here to be safer 🙂

shawn-ffufm commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -8,24 +10,38 @@ class PokemonOverviewPage extends StatelessWidget {
+        content: Text(errorMessage!),
+        duration: const Duration(seconds: 5),
+      );
+      key.currentState?.showSnackBar(snackBar);
```
seems no need for this, just use `ScaffoldMessenger.of(context).showSnackBar(SnackBar);`

shawn-ffufm commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -8,24 +10,38 @@ class PokemonOverviewPage extends StatelessWidget {
-        ),
-      ),
-    );
+        scaffoldMessengerKey: key,
```
seems no need for this

shawn-ffufm commented on lib/state/action/actions.dart:
```diff
@@ -0,0 +1,15 @@
+import 'package:async_redux/async_redux.dart';
+import 'package:pokedex_async_redux/state/app_state.dart';
+
+//Loading action
```
this is a comment not a documentation, it should be ///. Also the message doesn't tell anything what this action is for 😝

shawn-ffufm commented on lib/state/action/pokemon_actions.dart:
```diff
@@ -17,7 +22,7 @@ class GetPokemonsAction extends ReduxAction<AppState> {
 
 ///Getting of pokemon details from pokemon api
 class GetPokemonDetailsAction extends ReduxAction<AppState> {
-  GetPokemonDetailsAction({required this.pokemonName});
```
any reason why u've changed this one? 🤔

ChristianAsilo replied:
sorry po sir nagkamali lang po ako nag error po kasi then akala ko po is ireremove na based po doon sa pokemon_action nyo :sweat_smile:

shawn-ffufm replied:
any update on this one? 🙂

shawn-ffufm commented on lib/utils/constants.dart:
```diff
@@ -6,3 +8,9 @@ const pokemonName = 'charizard';
+//Pokemon Overview VM
+const errorMessage = 'Something went wrong!';
+const noPokemonAvailable = 'No pokemons available';
+final key = GlobalKey<ScaffoldMessengerState>();
```
see related feedback above about this 🙂

shawn-ffufm reviewed (changes requested):
see pr feedback 🙂

chrstn06 committed:
feat: Used union types for data, loading, and error
(pr feedback)
- Used null coalescing and emptyString
- Used ScaffoldMessenger.of(context).showSnackBar(snackBar);
- Modified documentation
- Removed unnecessary constants
- Removed the used of key

shawn-ffufm commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -8,24 +10,41 @@ class PokemonOverviewPage extends StatelessWidget {
-            return PokemonCard(
-              pokemon: pokemon,
+        appBar: AppBar(title: const Text(appbarTitle)),
+        body: Builder(
```
why wrap this in a Builder? seems no need for this u can maximize using the build method's provided context 🙂

shawn-ffufm commented on lib/state/action/pokemon_actions.dart:
```diff
@@ -17,7 +22,7 @@ class GetPokemonsAction extends ReduxAction<AppState> {
 ///Getting of pokemon details from pokemon api
 class GetPokemonDetailsAction extends ReduxAction<AppState> {
-  GetPokemonDetailsAction({required this.pokemonName});
+  GetPokemonDetailsAction(this.pokemonName);
```
see my feedback at the left side of this 🙂

shawn-ffufm reviewed (changes requested):
see minor pr feedback 🙂

chrstn06 committed:
feat: Used union types for data, loading, and error
(pr feedback)
- Modified pokemon_action to required
- Removed wrap in Builder
